### ISSUE TITLE
Using fairlearn with pytorch

### ISSUE THREAD

**Issue by rydeveraumn** [fairlearn/fairlearn#1341]
Hello all,

I am trying to utilize `fairlearn` with `pytorch` by utilizing the `skorch` library but the implementation has not been easy (at least in my opinion). Is there any examples or documentation to utilize the `fairlearn` library with `pytorch` models?

**Comment by hildeweerts** [comment]
Hi, @rydeveraumn! 

Unfortunately, we currently do not have examples or documentation including pytorch models (apart from our [adversarial mitigation user guide](https://fairlearn.org/v0.10/user_guide/mitigation/adversarial.html), but that one is related to a specific mitigation algorithm, rather than 'regular' use of fairlearn functionality with pytorch models).

However, please feel free to open an issue if you run into any particular problems!

**Comment by adrinjalali** [comment]
we don't have supporting `pytorch` out of the box in the roadmap. Using `skorch` on top of a `pytorch` model should be pretty straightforward, but if it's not, can you please open an issue on the `skorch` side, and also ping me there for me to see what the issues are? I can't necessarily fix the issues, but would like to see how `skorch` wants to handle it.

**Comment by rydeveraumn** [comment]
I am constrained on time so this may not be the best solution but it seems to work:

Here are the `skorch` modifications:

```python
class FairClassifier(NeuralNetBinaryClassifier):
    """
    Customized skorch wrapper build on top of
    NeuralNetBinaryClassifier to be utilized with
    the fairlearn library
    """

    def predict(self, X):
        predictions = self.predict_proba(X)
        predictions = (predictions > self.threshold).astype(int)
        return predictions

    def get_loss(self, y_pred, y_true, X=None, training=False):
        y_true = torch.tensor(y_true, dtype=torch.double, device=self.device)
        # Train condition
        train_sample_weight = self.module_.train_sample_weight
        test_sample_weight = self.module_.test_sample_weight
        train_condition = (
            train_sample_weight is not None
            and self.module_.training_samples == len(y_true)
        )
        test_condition = (
            test_sample_weight is not None
            and self.module_.testing_samples == len(y_true)
        )
        if train_condition:
            sample_weight = self.module_.train_sample_weight
        elif test_condition:
            sample_weight = self.module_.test_sample_weight
        else:
            sample_weight = None

        if sample_weight is not None:
            loss = self.criterion(y_pred, y_true)
            loss = loss * torch.tensor(sample_weight, dtype=torch.double)
            loss = loss.mean()
        else:
            loss = self.criterion_(y_pred, y_true).mean()

        return loss
```

and here are the modifications for a `torch` model:

```python
class FairlearnMLP(torch.nn.Module):
    """
    MLP to be utilized with fairlearn
    """

    def __init__(
        self,
        num_classes,
        num_features,
        num_layers=4,
        layer_width=128,
        out_activation='Sigmoid',
        training_samples=100,
        testing_samples=100,
        sensitive_group=None,
        bn=False,
    ):
        super(FairlearnMLP, self).__init__()
        # MLP model parameters
        layers = []
        for i in range(num_layers - 1):
            layers.append(
                nn.Linear(num_features if i == 0 else layer_width, layer_width)
            )
            if bn:
                layers.append(nn.BatchNorm1d(layer_width))
            layers.append(nn.ReLU())
        layers.append(nn.Linear(layer_width, num_classes))

        if out_activation == 'Sigmoid':
            layers.append(nn.Sigmoid())

        self.layers = nn.Sequential(*layers)

        # setup a dummy for sample weights
        self.train_sample_weight = None
        self.test_sample_weight = None
        self.training_samples = training_samples
        self.testing_samples = testing_samples
        self.sensitive_group = sensitive_group

    def forward(self, x, *, sample_weight=None):
        # Note: Had to use the below to get it to work with the aif360
        # fairness
        # data = kwargs
        # x = torch.hstack([v for k, v in data.items()])

        # If sample weights is not None save them so we can access
        # them in get_loss
        # train condition
        train_condition = sample_weight is not None and self.training_samples == len(
            sample_weight.values
        )

        # Test condition
        test_condition = sample_weight is not None and self.testing_samples == len(
            sample_weight.values
        )

        if train_condition:
            self.train_sample_weight = sample_weight.values
        elif test_condition:
            self.test_sample_weight = sample_weight.values

        x = self.layers(x)
        return x
```

The `fit` method in `sklearn` models have a method parameter called sample weight that adds weights to the loss function to help it learn differently than the regular loss. However, in the `torch` methods this parameter is not present and also does not happen in the `fit` method. So I added `sample_weight` but had the `get_loss` function handle the weights with `reduction='none'`. 

This is not the most elegant code but I had to find a way to handle the train and test data within these constraints so there is logic to flip between train and test weights (if there is a testing method involved). The way I am currently running is on smaller data and uses the full data length as a batch size.

I did have one more question about the `moments` - the constraint I am trying to emulate is:
$$\Bigg | \frac{1}{n_0} \sum_{i = 1}^{n_0} l(y_{i, 0}, g_{\theta}(\xi)) - \frac{1}{n_1} \sum_{i = 1}^{n_1} l(y_{i, 1}, g_{\theta}(\xi)) \Bigg | \leq r,$$

The closest moment I see for binary classification is the `EqualizedOdds` moment. Are there other moments that could emulate this constraint?

**Comment by adrinjalali** [comment]
@rydeveraumn this is more of a skorch issue than a fairlearn issue. Please post this there.
